docs: add examples for CephClient CRD use & libcephsqlite

t3hmrman opened the pull request:
**Description of your changes:**

I've added documentation that specifies how to use the `CephClient` CRD, along with how to use it specifically with the [Ceph SQLite VFS](https://docs.ceph.com/en/latest/rados/api/libcephsqlite/)

**Which issue is resolved by this Pull Request:**
Resolves #10689 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [x] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [x] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [x] Documentation has been updated, if necessary.
- [-] Unit tests have been added, if necessary.
- [-] Integration tests have been added, if necessary.

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+Extract Ceph cluster credentials from the generated secret (note that the subkey will be your original client name):
+
+```console
+kubectl get secret rook-ceph-client-example -o jsonpath="{.data.example}" -n rook-ceph | base64 -d
````
```suggestion
kubectl --namespace rook-ceph get secret rook-ceph-client-example -o jsonpath="{.data.example}" | base64 -d
```

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+To send writes to the cluster, you must retrieve the mons in use:
+
+```console
+kubectl get configmap rook-ceph-mon-endpoints -o jsonpath='{.data.data}' -n rook-ceph | sed 's/.=//g'`
````
```suggestion
kubectl --namespace rook-ceph get configmap rook-ceph-mon-endpoints -o jsonpath='{.data.data}' | sed 's/.=//g'`
```

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+Examples of the files follow:
+
+`ceph.conf`
+```toml
````
Does this render better as `ini`?

```suggestion
```ini
```

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+```
+
+`ceph.keyring`
+```toml
````
?

```suggestion
```ini
```

t3hmrman replied:
thanks!

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+  caps osd = 'profile rbd pool=<your pool>, profile rb pool=<another pool>'
+```
+
+`ceph.args`
+```toml
+--id example
+```
````
I'm having trouble understanding what the `ceph.args` file is. I assume this is meant to describe args you have to pass?

t3hmrman replied:
hey yes, this is for the use with `libcephsqlite` -- one of the things you can specify is the args to use with Ceph -- it's not necessary in the general case, so I can move it down to the`libcephsqlite` section probably

t3hmrman replied:
Actually thinking about it, I think it's necesary for `ceph` CLI use as well -- I guess I could change the command to `ceph status --id example` rather than specifying `CEPH_ARGS` via ENV.

What do you think?

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -41,3 +41,121 @@ spec:
+
+#### 5. Connect to the Ceph cluster with your given client ID
+
+Using your preferred method of choice (and/or created files), connect to Ceph with the 
```
Sentence trails off

t3hmrman replied:
- [ ] Fix incomplete sentence

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+
+The base64 encoded value that is returned *is* the password for your ceph client.
+
+#### 4. Determine the location of mons for the cluster
+
+To send writes to the cluster, you must retrieve the mons in use:
+
+```console
+kubectl get configmap rook-ceph-mon-endpoints -o jsonpath='{.data.data}' -n rook-ceph | sed 's/.=//g'`
+```
+
+This command *should* produce a line that looks somewhat like this:
+
+```
+XXX.XXX.XXX.XXX:6789,YYY.YYY.YYY.YYY:6789,ZZZ.ZZZ.ZZZ.ZZZ:6789
+```
````
This could get out of date over time, but that is not an issue with this documentation. Rather, perhaps @travisn and I should discuss making sure it is possible for apps in other k8s namespaces to reliably connect to the Ceph cluster directly when using the CephClient. There is a fair bit of manual work needed here that could be avoided with some additional feature work. 

I'm not sure if that should be allowing CephClients to be created in arbitrary namespaces, or if there are other solutions.

t3hmrman replied:
Yeah this is a great point, I look forward to seeing what ya'll decide. 

Originally that's what the issue I posted was about -- trying to get this manual work to zero with a more specific CRD to basically *do* this manual work. That said, at least having it in the documentation I think is great (and was a good learning experience for me), so you should have lots of time!

BlaineEXE commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,121 @@ spec:
+Apply your CephClient resource YAML (see above for an example client config) with `kubectl`:
+
+```console
+kubectl apply -f example.cephclient.yaml
````
Where is this example.cephclient.yaml file? It might be best to show the example using the `ceph-client.yaml` file that has cinder and glance users.

t3hmrman replied:
Oh I might need some feedback on that -- I don't think I made one that has *both* kinds of users. Will add this, and would appreciate some feedback!

t3hmrman replied:
Done! I added a new file with the example from the issue comments along with some explanatory comments.

BlaineEXE reviewed (commented):
This is a really great addition. Thank you very much. I have a few nits and questions, but nothing major.

t3hmrman commented:
Hey so I can't tell what is causing this, but any changes to `PendingReleaseNotes.md` are being reported as having a conflict. I tried to resolve via Github which resulted in a merge commit to my branch (yuck) and had to reverse that. 

If people are squashing all commits down to one anyway, why not fast-forward commits on `master`? Why have merge commits on master at all?

I rebased to `master` and addressed all the changes though -- so the PR *looks* right (with the new? change to `PendingReleaseNotes.md`).

t3hmrman commented:
Sorry ya'll, just realizing it's allowing only one reviewer at the same time so I guess @BlaineEXE you're first! :bow:

BlaineEXE commented:
I think you can remove changes from the pending release notes. We usually don't flag doc updates there anyway.

t3hmrman commented on PendingReleaseNotes.md:
```diff
@@ -8,3 +8,8 @@
 - The toolbox pod now uses the Ceph image directly instead of the Rook image
 - Add support for AWS Server Side Encryption with (AWS-SSE:S3)[https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingServerSideEncryption.html] for RGW.
+
+## Documentation
```
- [ ] remove pending release notes changes

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -41,3 +41,146 @@ spec:
 
 This guide assumes you have created a Rook cluster as explained in the main [Quickstart guide](../Getting-Started/quickstart.md).
+
+## Examples
```
- [ ] remove "examples" header

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -41,3 +41,146 @@ spec:
 ### Prerequisites
 
 This guide assumes you have created a Rook cluster as explained in the main [Quickstart guide](../Getting-Started/quickstart.md).
```
- [ ] move prereq section to before creating ceph user

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -41,3 +41,146 @@ spec:
+
+To use `CephClient` to connect to a Ceph cluster:
+
+#### 1. Create the `CephClient` (`metadata.name` will be your Ceph client ID)
```
- [ ] remove added ceph client creation in favor of existing client section

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -41,3 +41,146 @@ spec:
+This command *should* produce a line that looks somewhat like this:
+
+```
+XXX.XXX.XXX.XXX:6789,YYY.YYY.YYY.YYY:6789,ZZZ.ZZZ.ZZZ.ZZZ:6789
````
- [ ] use example IP addresses instead of placeholders

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -41,3 +41,146 @@ spec:
+
+## Examples
+
+### Using a `CephClient` to create Ceph credentials
```
- [ ] remove header about using `CephClient` to create creds

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -34,10 +38,121 @@ metadata:
+Once your `CephClient` has been processed by Rook, it will be updated to include your secret:
+
+```console
+kubectl -n rook-ceph get cephclient example -o jsonpath='{.status.info.secretName}'
````
Ah yes, you're right -- I'll change it there too

t3hmrman replied:
ended up moving this back to `example` because using the same client for both examples makes a bit less sense naming wise (`sqlitevfs` means nothing to someone that just wants the first section with no intent to use it for the SQLite VFS)

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -5,14 +5,18 @@ title: Client CRD
+
+### 1. Creating Ceph User
 
 To get you started, here is a simple example of a CRD to configure a Ceph client with capabilities.
```
These CephClient CRs are from the previous documentation -- I'm not sure what the intent for use for the previous ones was... It seems like a profile that works with OpenStack Cinder and Glance? I don't use OpenStack so I'm not sure if someone who does would have found that config useful.

Is it fine that I delete them and replace with the one for this specific case?

t3hmrman replied:
OK so it doesn't really work to have `sqlitevfs` be the name for an example case -- so I'll switch it back to `example` everywhere, and replace the existing clients.

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -34,10 +38,121 @@ metadata:
+  caps osd = 'profile rbd pool=<your pool>, profile rb pool=<another pool>'
+```
+
+`ceph.args`
````
yeah I personally like the file, but since I went with inlining it I'll remove the file.

People can use a file and read contents if they find it easier

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -34,10 +38,121 @@ metadata:
+```console
+export CEPH_CONF=/libsqliteceph/ceph.conf;
+export CEPH_KEYRING=/libsqliteceph/ceph.keyring;
+export CEPH_ARGS=`cat /libsqliteceph/ceph.args`;
````
- [ ] set ceph args directly in final instructions

t3hmrman commented on deploy/examples/sqlitevfs-client.yaml:
```diff
@@ -0,0 +1,182 @@
+              echo -e "[info] writing out ceph.conf...";
+              echo -e "[global]" > /libsqliteceph/ceph.conf;
+              echo -e "mon_host=${MON_ENDPOINTS}" >> /libsqliteceph/ceph.conf;
+              echo -e "#debug ms = 1/6" >> /libsqliteceph/ceph.conf;
```
- [ ] remove commented out lines in ceph.conf

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -34,10 +38,121 @@ metadata:
+Extract Ceph cluster credentials from the generated secret (note that the subkey will be your original client name):
+
+```console
+kubectl --namespace rook-ceph get secret rook-ceph-client-example -o jsonpath="{.data.example}" | base64 -d
````
it would be `rook-ceph-client-sqlitevfs`, will change it to that!

- [ ] change `rook-ceph-client-example` to `rook-ceph-client-sqlitevfs`

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
````diff
@@ -5,39 +5,139 @@ title: Client CRD
+```console
+export CEPH_CONF=/libsqliteceph/ceph.conf;
+export CEPH_KEYRING=/libsqliteceph/ceph.keyring;
+export CEPH_ARGS=--id example
````
Sure!

- [ ] rename ceph args client in sqlite example

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -5,39 +5,139 @@ title: Client CRD
+
+Without the appropriate package (or a from-scratch build of SQLite), you will be unable to load `libcephsqlite.so`.
+
+Assuming you have already *successfully retrieved and set up Ceph client credentials* using `CephClient`, you may use those credentials with SQLite by running the following:
```
- [ ] link to example client file from sqlite example

t3hmrman commented on deploy/examples/sqlitevfs-client.yaml:
```diff
@@ -0,0 +1,178 @@
+
+              export CEPH_CONF=/libsqliteceph/ceph.conf;
+              export CEPH_KEYRING=/libsqliteceph/ceph.keyring;
+              export CEPH_ARGS=`cat /libsqliteceph/ceph.args`;
```
Oh sorry this got left in there!

Thank you for the thorough review :bow:

t3hmrman replied:
- [ ] remove args file use

t3hmrman commented on deploy/examples/sqlitevfs-client.yaml:
````diff
@@ -0,0 +1,178 @@
+# ```
+# export CEPH_CONF=/libsqliteceph/ceph.conf;
+# export CEPH_KEYRING=/libsqliteceph/ceph.keyring;
+# export CEPH_ARGS=`cat /libsqliteceph/ceph.args`;
````
- [ ] remove args file use

t3hmrman commented on Documentation/CRDs/ceph-client-crd.md:
```diff
@@ -5,39 +5,139 @@ title: Client CRD
+
+Without the appropriate package (or a from-scratch build of SQLite), you will be unable to load `libcephsqlite.so`.
+
+After creating a `CephClient` similar to [`deploy/examples/sqlitevfs-client.yaml`](../../deploy/examples/sqlitevfs-client.yaml) and retrieving it's credentials, you may set the following ENV variables:
```
- [ ] fix link to `sqlitevfs-client.yaml`

t3hmrman commented:
Thanks again for the patient review! :bow:
